File: src/Set3a.hs
{-# LANGUAGE NoImplicitPrelude #-}

module Set3a where

import           MCPrelude

allPairs :: [a] -> [b] -> [(a, b)]
allPairs [] _                  = []
allPairs _ []                  = []
allPairs as@(a:as') bs@(b:bs') = (a, b) : allPairs [a] bs' ++ allPairs as' bs

allPairs' :: [a] -> [b] -> [(a, b)]
allPairs' as bs = concat $ map (\a -> map (\b -> (a, b)) bs) as

allPairs'' :: [a] -> [b] -> [(a, b)]
allPairs'' as bs = foldl (\acc a -> acc ++ (map (\b -> (a, b)) bs)) [] as

-- This doesn't work (!!!) pattern-matching fails
-- but you can do this in JS/TS
allPairs''' :: [a] -> [b] -> [(a, b)]
allPairs''' xs@(x:xs') ys@(y:ys')
  | null xs = []
  | null ys = []
  | otherwise = (x, y) : allPairs''' [x] ys' ++ allPairs''' xs' ys

File: src/typescript/set3a.ts
/*
allPairs :: [a] -> [b] -> [(a, b)]
allPairs [] _                  = []
allPairs _ []                  = []
allPairs xs@(x:xs') ys@(y:ys') = (x, y) : allPairs [x] ys' ++ allPairs xs' ys
*/
export const allPairs = <A>(xs: Array<A>, [x, ...xs_] = xs) => <B>(
  ys: Array<B>,
  [y, ...ys_] = ys
): Array<[A, B]> =>
  xs.length === 0
    ? []
    : ys.length === 0
    ? []
    : [[x, y], ...allPairs([x])(ys_), ...allPairs(xs_)(ys)];

allPairs([1, 2, 3])([4, 5, 6]); //?

/*
allPairs' :: [a] -> [b] -> [(a, b)]
allPairs' as bs = foldl (\acc a -> acc ++ (map (\b -> (a, b)) bs)) [] as
*/
export const allPairs_ = <A>(as: Array<A>) => <B>(
  bs: Array<B>
): Array<[A, B]> => as.reduce((acc, a) => [...acc, ...bs.map(b => [a, b])], []);
